### ISSUE TITLE
A small fix for Matrix Export

### ISSUE THREAD

**Issue by msdogan** [ucd-cws/calvin-network-tools#35]
@jrmerz 
@jdherman identified that lower bound information is completely missing for five reservoirs for `FINAL` storage link. For other months, it is fine. This could be because that 'endingstorage' values for these reservoirs were updated as 'null'. Upper bounds are correct but lower bound values should be '0' for all of them for `FINAL` storage. Thanks.

```
i   j   k   cost    amplitude   lower_bound upper_bound
SR_TL.2003-09-30    FINAL   0   0   1       10000000
SR_BVLB.2003-09-30  FINAL   0   0   1       10000000
SR_OL.2003-09-30    FINAL   0   0   1       10000000
SR_ML.2003-09-30    FINAL   0   0   1       10000000
SR_SS.2003-09-30    FINAL   0   0   1       10000000

```

**Comment by msdogan** [comment]
@jrmerz there is actually something interesting going on for those links. If I open csv file with Excel after exporting, numbers are not there, as shown above. However, if I open the same file with notepad++ or sublime text, lower bound numbers show up as 0, which is correct. Any idea why?

**Comment by jdherman** [comment]
That is strange. This looks like tsv not csv, so can Excel read it correctly? Is Excel changing the file? (are the numbers still there if you open with a text editor first?)

The text editors I trust ... Excel might be changing things up by making some assumptions about delimiters or missing values.

Were the versions of the *dat file with errors created using Excel and then copy+pasted, or did they come directly from the matrix export? Maybe there's an intermediate step that's causing the problem.

**Comment by jrmerz** [comment]
@msdogan yes, you normally export as tsv, not csv and that could be tripping up Excel.  What is the extension name on the file?  I don't have access to Excel, but look around there should be a setting to manually set the delimiter.  Or open in your favorite text editor and replace all tabs '\t' with spaces '\s' (you can also set the export delimiter from the cnf matrix flag --fs).

**Comment by msdogan** [comment]
@jdherman actually that was a csv file. I opened that with Excel that's why you don't see commas. If you open the same file with other text editors, commas are there. Link for SR_SS is shown below. First one is opened with Excel and second with Sublime Text. Lower bound numbers are not there. Not sure how magically appeared yesterday when I tried with Sublime text, but final decision is lb values are missing for those five reserboirs.

![excel](https://cloud.githubusercontent.com/assets/14085662/17063694/6de42aae-4fed-11e6-9276-50870c99d2f5.JPG)

![subtext](https://cloud.githubusercontent.com/assets/14085662/17063698/70bdadb8-4fed-11e6-9779-5e8f96671b62.JPG)

**Comment by msdogan** [comment]
@jrmerz I use this command. does is not export as csv?

```
 cnf matrix --format csv --start 2002-10 --stop 2003-10 --ts . --to networklinks --max-ub 10000000 --outnodes networknodes --debug All
```

**Comment by msdogan** [comment]
Last two files are output of the command.

![capture](https://cloud.githubusercontent.com/assets/14085662/17063864/4796cb80-4fee-11e6-9470-53c18293580d.JPG)

**Comment by jrmerz** [comment]
The --format csv flag means the general '_____ Separator Value' format, with commas (CSV) being the most common the TSV option in the docs is misleading.  The flag is really to select between _SV, PNG or DOT formats.  Then you can supply the delimiter for the file with the --ts flag.

The command you provided looks like it should actually be using '.' as the delimiter...  Is that not the case?

**Comment by jrmerz** [comment]
Can you turn on file extensions?  I believe we stamp all _SV files with CSV, though TSV is more correct for tabs.

**Comment by msdogan** [comment]
@jrmerz I think we need tsv (tab separated values, right?) data file as a input for Pyomo. 

Is this what you are looking for?

![capture](https://cloud.githubusercontent.com/assets/14085662/17063953/c771cd3c-4fee-11e6-8a7c-2b6051ca2772.JPG)

**Comment by msdogan** [comment]
@jrmerz if I want to use 'tab' as delimiter, what would the command look like?

what do I enter after `--fs` command? just `tab` ?

```
cnf matrix --start 2002-10 --stop 2003-10 --ts . --to networklinks --max-ub 10000000 --outnodes networknodes --debug All
```

**Comment by jrmerz** [comment]
So, yes, I miss read your command above.  Before, you would set `--format csv --fs :tab:`  then rename the file to tsv so excel wasn't confused (and to be more formally correct).

However, I just pushed v1.0.42 which now formally support `--format tsv`.  No --fs or file rename required.  Give that a go.

**Comment by msdogan** [comment]
@jrmerz this is exactly what I am looking for :) thanks a lot. However, lower bound values are still missing. If you can solve, I will not bug you :) Except there is also sink issue, which is very important.

**Comment by jrmerz** [comment]
@msdogan Looks like the current endingstorage is null for SR_SS (https://raw.githubusercontent.com/ucd-cws/calvin-network-data/master/data/colorado-river/imperial-valley/sr_ss/node.geojson).  You set it to null on 7/15 https://github.com/ucd-cws/calvin-network-data/commit/cc1a5addc3df4c9f0982f910150e03bf733bb08e

And since the `--stop 2003-10` is passed the time series storage values (2003-09-30), the ending storage is used.

**Comment by msdogan** [comment]
@jrmerz that's is so correct. we want no ending storage requirement for `SR_SS, SR_ML, SR_OL, SR_TL and SR_BVLB` . I would expect to see lb as 0 and ub as 1000000, which I define while exporting. Can we define another flag, like if lower bound is null then set to zero. This could also be default. right @jdherman? We don't want them to go below zero.

Edit: Just to clarify, for those reservoirs, no ending storage requirements only for very last time-step, which is September 2003. If you create a run with smaller time-step or if final period is different than September 2003, then yeah it is not null anymore and should be constrained, just like we do for .pri export.

**Comment by jrmerz** [comment]
@msdogan @qjhart bringing in quinn.

Just let me know, I am following orders on this one. :)

**Comment by jdherman** [comment]
@msdogan if you want no ending storage requirement, then the bounds should be from 0 to (arbitrary large number). So yes I agree that the null's should become 0 lb in the matrix export. 

Edit: about the tsv/csv issue above, it's no problem if we export the matrix in CSV and then convert it later as a preprocessing step. CSV is more general, just pyomo for whatever reason expects tsv.

**Comment by jrmerz** [comment]
So is that a go for this change?

**Comment by jdherman** [comment]
Double check with Quinn, but I vote yes. Thanks for your help with this :)

**Comment by jrmerz** [comment]
fixed push calvin-network-tools@1.0.45

**Comment by msdogan** [comment]
For final storage values for selected natural lakes appears to have 0 lower bound as default for the time-period of Sep 2003. So, it works. Closing the issue.
